### ISSUE TITLE
To let or not to let

### ISSUE THREAD

**Issue by JacquesCarette** [agda/agda-stdlib#2369]
Right now, `let` in Agda generates substitutions (as confirmed by Andreas [here](https://github.com/agda/agda/issues/5497) amongst other places). The [attempted fix](https://github.com/agda/agda/issues/3094) for this has stalled even though it is an [extremely old](https://github.com/agda/agda/issues/514) issue.

The advice is to favour `where` over `let`.

The question is: do we program stdlib to current Agda or some optimistic future Agda that will have a better implementation of `let`?

Given just how long this has been a wart, I'm of the opinion that being optimistic would be ill-placed (even though I'm actually normally fairly optimistic!)

**Comment by gallais** [comment]
FTR, these cannot be turned into `where` definitions because these get lifted to the
toplevel and that is not done in a modality-respecting way. In other words: turning
these into `where` definitions leads to errors about irrelevant values being used in
a relevant context.

https://github.com/gallais/agda-stdlib/blob/system-random/src/System/Random.agda#L119-L133

**Comment by JacquesCarette** [comment]
Good point: indeed there are times when we have no choice, because the implementation scheme for `where` is also sub-optimal.

**Comment by jamesmckinna** [comment]
Time to revisit after 2.7.0's introduction of `using`?

**Comment by JacquesCarette** [comment]
That would be great - but we'd need to know how `using` is implemented to see if it's actually a win. Perhaps @jespercockx can enlighten us?

**Comment by jespercockx** [comment]
I was not the one who implemented it, but I believe `using` is desugared to an auxiliary function, so I'm afraid it's not really better performance-wise than a `where` declaration.

I started working on implementing proper `let`-bindings for Agda's internal syntax during the last Agda meeting (see https://github.com/agda/agda/pull/7253) but it's a lot more complicated that I had anticipated so don't expect anything from it soon.
